core/Symbols.h

```diff
@@ -623,6 +654,9 @@ class Symbol final {
      */
     InlinedVector<SymbolRef, 4> mixins_;
 
+    // Ancestors required by this class or module associated to the `requires_ancestor X` location
+    InlinedVector<Symbol::RequiredAncestor, 4> requiredAncestorsLocs;
```

Comment:
This is a ton of memory to dedicate to something which is almost never used. This is going to make the Symbol table much, much larger. `RequiredAncestor` is 4 bytes, and you're reserving 4 of them inline.

In general, I think you're storing too much information. We don't store a `Loc` nor a `requiredBy` `SymbolRef` for `mixins_`. Why do we need to for `requiredAncestors`? Don't you have enough information to compute that information at the error site, like we do for `mixins_`? I think we should only need to store the `SymbolRef required;` field, which would mean we could delete the `RequiredAncestor` struct entirely.

Reply:
On top of that, I would probably drop this down to `InlinedVector<..., 1>` because that's more likely to be the common case.

The way we found the numbers for every other `InlinedVector` in Sorbet before was to instrument Sorbet to dump a bunch of counters, and then burn those numbers in.

But for a new feature which is almost never used, it's not so much required. Maybe after a year of usage we could revisit it but a default of `1` is probably the best bet (most people are just going to do `Kernel` and call it a day).

Reply:
* `requiredBy` is used to displayed which module required the ancestor (it might not be `self`). It's super useful to debug:

    ```
    B is missing X (required by Y)
    ```

    It's cached to make the error reporting faster but for this information I think I can go up the hierarchy and find the source.

* `requiredAt` is used to display the proper location of where `X` was required in `Y`. I believe that after the `resolver` this information is lost for ever. I think I used the same approach than for the `sealedLocs`: https://github.com/sorbet/sorbet/blob/master/core/Symbols.h#L522

We can limit the memory footprint but I fear it will be for less expressive errors.

Reply:
I'm not convinced. We emit pretty much the same errors in `definition_validator` but we don't track the `Loc` information there.

In fact, did you ever look at implementing as much as possible of these error-emitting checks in `definition_validator`? Right now we're doing a huge loop over every `Symbol` in `GlobalState`. If these checks were in `definition_validator`, they would only run for the files that changed.

Reply:
The whole thing is heavily inspired from the validator for `abstract` (you can see it with the caching mechanism actually). At first I put the checks in the `definition_validator` but I also need to change the GlobalPass to perform the linearization of the required ancestors (so method resolution works). So I moved everything to the class in the `GlobalPass` to minimize the diffs.